### PR TITLE
LoggingConfiguration - Initialize should warn when LogManager.ThrowExceptions=true

### PR DESCRIPTION
**Pull request by snakefoot**
LogManager.ThrowExceptions=true can give you a bad skin rash, hair-loss and bleeding headache, along with other severe side-effects.

### CODE REVIEW

File: src/NLog/Config/LoggingConfiguration.cs [hunk 1]
```diff
@@ -761,6 +761,11 @@ internal void InitializeAll()
         {
             bool firstInitializeAll = _configItems.Count == 0;
 
+            if (firstInitializeAll && (LogFactory.ThrowExceptions || LogManager.ThrowExceptions))
+            {
+                InternalLogger.Info("LogManager.ThrowExceptions = true will have severe side-effects. Use only for unit-testing and last resort troubleshooting.");
```
Comment (304NotModified) [inline src/NLog/Config/LoggingConfiguration.cs]: Good idea!

The warning sounds a bit scary. 

Proposal: 

> LogManager.ThrowExceptions  is enabled. It's recommend to disable LogManager.ThrowExceptions in production environments. See https://github.com/nlog/nlog/wiki/SOMEPAGE.

**Comment by snakefoot** [comment]
The warning is real. You should not use it. Will make NLog fail or stall and possible take down the entire application. Not sure that it is a good thing.


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


@304NotModified commented on this pull request.

________________________________

In src/NLog/Config/LoggingConfiguration.cs<https://github.com/NLog/NLog/pull/3773#discussion_r371028932>:

> @@ -761,6 +761,11 @@ internal void InitializeAll()
         {
             bool firstInitializeAll = _configItems.Count == 0;

+            if (firstInitializeAll && (LogFactory.ThrowExceptions || LogManager.ThrowExceptions))
+            {
+                InternalLogger.Info("LogManager.ThrowExceptions = true will have severe side-effects. Use only for unit-testing and last resort troubleshooting.");


Good idea!

The warning sounds a bit scary.

Proposal:

LogManager.ThrowExceptions is enabled. It's recommend to disable LogManager.ThrowExceptions in production environments. See https://github.com/nlog/nlog/wiki/SOMEPAGE.

**Comment by 304NotModified** [comment]
"severe side-effects" sounds like corrupt machines, IIS resets and open security holes. ;)

**Comment by snakefoot** [comment]
> "severe side-effects" sounds like corrupt machines, IIS resets and open security holes. ;)

I guess denial-of-service attack because service is stuck some where nasty inside NLog is some kind of corruption.

**Comment by 304NotModified** [comment]
maybe, but the DOS possibility is not only controlled by LogManager.ThrowExceptions.

**Comment by snakefoot** [comment]
> maybe, but the DOS possibility is not only controlled by LogManager.ThrowExceptions.

But `LogManager.ThrowExceptions` makes it much more likely that NLog will explode in your face, and leave you hanging.

**Comment by 304NotModified** [comment]
new proposal:

> LogManager.ThrowExceptions is enabled.  As LogManager.ThrowExceptions could take down your application, it's recommend to set LogManager.ThrowExceptions to false. See https://github.com/nlog/nlog/wiki/SOMEPAGE.

**Comment by snakefoot** [comment]
Changed to:

> LogManager.ThrowExceptions = true can cause application crashes. Use only for unit-testing and last resort troubleshooting.

**Comment by snakefoot** [comment]
Changed to:

> LogManager.ThrowExceptions = true can crash application! Use only for unit-testing and last resort troubleshooting.

**Comment by 304NotModified** [comment]
OK merged :)

**Pull request merged**